woocommerce/woocommerce-gateway-stripe
# Find existing refunds by refund ID when processing webhook

Fixes #2580 

## Changes proposed in this Pull Request:



When handling a refund webhook, this attempts to lookup any existing refund by the refund ID first, before falling back to the current approach using the charge ID. As described in #2580, this will allow other extensions to decouple refunds from the original order the charge is against. The current approach leads to duplicate refunds on the WooCommerce side in these situations. 

In general it seems to be more intuitive and more reliable to lookup the refund directly by the refund ID like this, rather than going via the order with the same charge ID.



## Testing instructions



The basic behaviour to be tested:
- Refunds initiated from WooCommerce result in webhook, which bails before creating a new refund, due to the presence of the existing refund
- Refunds initiated from Stripe admin result in webhook, which creates a new refund within WooCommerce

In addition, to test the scenario I'm hoping to support:
- When two orders are given the same charge ID (for testing this can just be done directly in the database), refunds initiated from one order in WooCommerce do not add a duplicate refund on the other order

The basic behaviour should not be changed so existing e2e tests for refunds should continue to be relevant. As far as I can tell there's nothing to specifically test for duplicate refunds though.



---


**Post merge**
